jitbuilder/lljb/src/IRVisitor.cpp

```diff
@@ -0,0 +1,559 @@
+   _methodBuilder->mapIRtoIlValue(&I, result);
+   }
+
+void
+IRVisitor::visitCmpInst(llvm::CmpInst &I)
+   {
+   TR::IlValue * lhs = getIlValue(I.getOperand(0));
+   TR::IlValue * rhs = getIlValue(I.getOperand(1));
+
+   TR::IlValue * result = nullptr;
+
+   switch (I.getPredicate())
+      {
+      case llvm::CmpInst::Predicate::ICMP_EQ:
+      case llvm::CmpInst::Predicate::FCMP_OEQ:
+      case llvm::CmpInst::Predicate::FCMP_UEQ:
+         result = _builder->EqualTo(lhs, rhs);
+         break;
+      case llvm::CmpInst::Predicate::ICMP_NE:
+      case llvm::CmpInst::Predicate::FCMP_ONE:
+      case llvm::CmpInst::Predicate::FCMP_UNE:
+         result = _builder->NotEqualTo(lhs, rhs);
+         break;
+      case llvm::CmpInst::Predicate::ICMP_UGT:
+         result = _builder->UnsignedGreaterThan(lhs, rhs);
+         break;
+      case llvm::CmpInst::Predicate::ICMP_UGE:
+         result = _builder->UnsignedGreaterOrEqualTo(lhs, rhs);
+         break;
+      case llvm::CmpInst::Predicate::ICMP_ULT:
+         result = _builder->UnsignedLessThan(lhs, rhs);
+         break;
+      case llvm::CmpInst::Predicate::ICMP_ULE:
+         result = _builder->UnsignedLessOrEqualTo(lhs, rhs);
+         break;
+      case llvm::CmpInst::Predicate::ICMP_SGT:
+      case llvm::CmpInst::Predicate::FCMP_OGT:
+      case llvm::CmpInst::Predicate::FCMP_UGT:
+         result = _builder->GreaterThan(lhs, rhs);
+         break;
+      case llvm::CmpInst::Predicate::ICMP_SGE:
+      case llvm::CmpInst::Predicate::FCMP_OGE:
+      case llvm::CmpInst::Predicate::FCMP_UGE:
+         result = _builder->GreaterOrEqualTo(lhs, rhs);
+         break;
+      case llvm::CmpInst::Predicate::ICMP_SLT:
+      case llvm::CmpInst::Predicate::FCMP_OLT:
+      case llvm::CmpInst::Predicate::FCMP_ULT:
+         result = _builder->LessThan(lhs, rhs);
+         break;
+      case llvm::CmpInst::Predicate::ICMP_SLE:
+      case llvm::CmpInst::Predicate::FCMP_OLE:
+      case llvm::CmpInst::Predicate::FCMP_ULE:
+         result = _builder->LessOrEqualTo(lhs, rhs);
+         break;
+      default:
+         assert(0 && "Unknown CmpInst predicate");
+         break;
+      }
+   _methodBuilder->mapIRtoIlValue(&I, result);
+   }
+
+void
+IRVisitor::visitBranchInst(llvm::BranchInst &I)
+   {
+   if (I.isUnconditional())
+      {
+      llvm::Value * dest = I.getSuccessor(0);
+      TR::BytecodeBuilder * destBuilder = _methodBuilder->getByteCodeBuilder(dest);
+      assert(destBuilder && "failed to find builder for target basic block");
+      _builder->Goto(destBuilder);
+      }
+   else
+      {
+      TR::IlValue * condition = getIlValue(I.getCondition());
+      TR::IlBuilder * ifTrue = _methodBuilder->getByteCodeBuilder(I.getSuccessor(0));
+      TR::IlBuilder * ifFalse = _methodBuilder->getByteCodeBuilder(I.getSuccessor(1));
+      assert(ifTrue && ifFalse && condition && "Failed to find destination blocks for ifThenElse");
+      _builder->IfThenElse(&ifTrue, &ifFalse, condition);
```

Comment:
I do not think this is what you want, although you might be getting away with it by fluke. I think you would be better off doing:
```
_builder->IfCmpEqualTo(&ifTrue, _builder->NotEqual(condition, _builder->Const*(0)));
_builder->Goto(&ifiFalse);
```

(not sure what kind of Const you need).

Reply:
Makes sense, thanks for pointing this out. I have incorporated your suggestion here.